### PR TITLE
Wait on close

### PR DESCRIPTION
**Pull request by ohadbitt**
#### Pull Request Description
The sink task stop and close should first stop timers from ingesting anymore data so that we don't create duplicates when topic is opened again.
Resolving issue https://github.com/Azure/kafka-sink-azure-kusto/issues/69
---

**Fixes:**
- Possible duplicates on sink task stop

### CODE REVIEW

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java [hunk 1]
```diff
@@ -181,15 +193,20 @@ public void rollback() throws IOException {
...
         // Flush last file, updating index
         finishFile(true);
+    }
+
+    public void stop() {
+        stopped = true;
 
-        // Setting to null so subsequent calls to close won't write it again
-        currentFile = null;
+        if (timer != null) {
```
Comment (AsafMah) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:205]: If this is called in parallel - timer could turn into null between the check and use, and you'd have an NPE

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java [hunk 2]
```diff
@@ -160,15 +168,19 @@ private void handleErrors(String message, Exception e) {
...
+    // Synchronize for Can be called from rollback - in that
     private void dumpFile() throws IOException {
-        countingStream.close();
-        currentFileDescriptor = null;
-        boolean deleted = currentFile.file.delete();
-        if (!deleted) {
-            log.warn("couldn't delete temporary file. File exists: " + currentFile.file.exists());
+        SourceFile temp = this.currentFile;
+        if (temp != null) {
+            countingStream.close();
```
Comment (AsafMah) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:176]: No temp for this?

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java [hunk 3]
```diff
@@ -219,21 +236,24 @@ void flushByTimeImpl() {
...
+
             resetFlushTimer(false);
-            reentrantReadWriteLock.writeLock().unlock();
         } catch (Exception e) {
             String fileName = currentFile == null ? "no file created yet" : currentFile.file.getName();
             long currentSize = currentFile == null ? 0 : currentFile.rawBytes;
             flushError = String.format("Error in flushByTime. Current file: %s, size: %d. ", fileName, currentSize);
             log.error(flushError, e);
+        } finally {
+            reentrantReadWriteLock.writeLock().unlock();
```
Comment (AsafMah) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:252]: If we'll merge https://github.com/Azure/kafka-sink-azure-kusto/pull/67 you can use the auto lock

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java (hunk 2 above)
Comment (ohadbitt) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:176]: ?

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java (hunk 3 above)
Comment (ohadbitt) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:252]: ok

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java (hunk 1 above)
Comment (AsafMah) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:201]: How can you be sure?  
Remember that you also call it from TopicPartitionWriter.stop and TopicPartitionWriter.close. Can these also not be called in parallel? and what about what calls them?

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java [hunk 4]
```diff
@@ -162,13 +169,16 @@ private void handleErrors(String message, Exception e) {
...
-        countingStream.close();
-        currentFileDescriptor = null;
-        boolean deleted = currentFile.file.delete();
-        if (!deleted) {
-            log.warn("couldn't delete temporary file. File exists: " + currentFile.file.exists());
-        }
+        SourceFile temp = currentFile;
         currentFile = null;
+        if (temp != null) {
+            countingStream.close();
```
Comment (AsafMah) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:175]: This is called from rollback()
rollback() already closes `countingStream` (line 186). Isn't it bad to double close a stream?

Comment (ohadbitt) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:175]: no

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java (hunk 1 above)
Comment (ohadbitt) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/FileWriter.java:201]: i guess i can never be sure but the logic is - some kafka thread accepts a command to lose partition - then it calls stop here . So unless two such commands are sent somehow it shouldnt happen

**Pull request merged**